Content.Shared/RCD/Systems/RCDSystem.cs

```diff
@@ -35,313 +41,590 @@ public sealed class RCDSystem : EntitySystem
+
+            return false;
+        }
+
+        // Check rule: Must place on subfloor
+        if (component.CachedPrototype.ConstructionRules.Contains(RcdConstructionRule.MustBuildOnSubfloor) && !mapGridData.Tile.Tile.GetContentTileDefinition().IsSubFloor)
+        {
+            if (popMsgs)
+                _popup.PopupClient(Loc.GetString("rcd-component-cannot-build-as-subfloor-not-exposed-message"), uid, user);
 
-        if (!unobstructed)
             return false;
+        }
 
-        switch (comp.Mode)
+        // Tile specific rules
+        if (component.CachedPrototype.Mode == RcdMode.ConstructTile)
         {
-            //Floor mode just needs the tile to be a space tile (subFloor)
-            case RcdMode.Floors:
-                if (!tile.Tile.IsEmpty)
+            // Check rule: Tile placement is valid
+            if (!_floors.CanPlaceTile(mapGridData.GridUid, mapGridData.Component, out var reason))
+            {
+                if (popMsgs)
+                    _popup.PopupClient(reason, uid, user);
+
+                return false;
+            }
+
+            // Check rule: Tiles can't be identical
+            if (mapGridData.Tile.Tile.GetContentTileDefinition().ID == component.CachedPrototype.Prototype)
+            {
+                if (popMsgs)
+                    _popup.PopupClient(Loc.GetString("rcd-component-cannot-build-as-tiles-are-identical"), uid, user);
+
+                return false;
+            }
+
+            // Ensure that all construction rules shared between tiles and object are checked before exiting here
+            return true;
+        }
+
+        // Object specific rules
+
+        // Check rule: The tile is unoccupied
+        var isWindow = component.CachedPrototype.ConstructionRules.Contains(RcdConstructionRule.IsWindow);
+        var constructionAngle = (component.CachedPrototype.Rotation == RcdRotation.User) ? component.ConstructionDirection.ToAngle() : Direction.South.ToAngle();
+
+        foreach (var ent in _lookup.GetLocalEntitiesIntersecting(mapGridData.Tile, -0.05f, LookupFlags.Uncontained))
+        {
+            if (isWindow && HasComp<SharedCanBuildWindowOnTopComponent>(ent))
+                continue;
+
+            if (component.CachedPrototype.CollisionMask != CollisionGroup.None && TryComp<FixturesComponent>(ent, out var fixtures))
+            {
+                for (int i = 0; i < fixtures.FixtureCount; i++)
                 {
-                    _popup.PopupClient(Loc.GetString("rcd-component-cannot-build-floor-tile-not-empty-message"), uid, user);
+                    (var _, var fixture) = fixtures.Fixtures.ElementAt(i);
+
+                    // No collision possible
+                    if (Prototype(ent)?.ID != component.CachedPrototype.Prototype &&
+                        (fixture.CollisionLayer <= 0 || (fixture.CollisionLayer & (int) component.CachedPrototype.CollisionMask) == 0))
```

Comment:
You can't use prototype IDs to compare entities as it's just used for spawn logic and no guarantee of what an entity is.